extensions/wikia/FilePage/WikiaFilePage.php

```diff
@@ -40,29 +40,105 @@ protected function isVideo() {
 	/**
 	 * Render the image or video
 	 */
-	protected function openShowImage(){
+	protected function openShowImage() {
 		if ( $this->isVideo() ) {
 			$this->openShowVideo();
 		} else {
 			parent::openShowImage();
 		}
 	}
 
+	/**
+	 * Render the image or video
+	 */
+	public function view() {
+		global $wgMemc, $wgDBprefix;
+		if ( !$this->getContext()->getUser()->isAnon() ) {
+			parent::view();
+
+			return;
+		}
+		//fallback to main page
+		$url = Title::newMainPage()->getFullURL();
+		//wiki needs read privileges
+		if ( !$this->getTitle()->userCan( 'read' ) ) {
+			$this->getContext()->getOutput()->redirect( $url );
+
+			return;
+		}
+		$redirKey = wfMemcKey( "redir:", $wgDBprefix, $this->getTitle()->getPrefixedText() );
+
+		$url = $wgMemc->get( $redirKey );
+		if ( $url ) {
+			$this->getContext()->getOutput()->redirect( $url );
+
+			return;
+		}
+		$displayImg = $img = false;
+		Hooks::run( 'ImagePageFindFile', [ $this, &$img, &$displayImg ] );
+		if ( !$img ) { // not set by hook?
+			$img = wfFindFile( $this->getTitle() );
+			if ( !$img ) {
+				$img = wfLocalFile( $this->getTitle() );
+			}
+		}
+		if ( !$img ) {
+			parent::view();
+
+			return;
+		}
+		$res = $this->fetchLinks( $img->getTitle()->getDBkey() );
+
+		foreach ( $res as $row ) {
+			$title = Title::newFromRow( $row );
+			if ( $title->isRedirect() ) {
+				continue;
+			}
+			$url = $title->getFullURL();
+			$res->free();
+			break;
+		}
+		$wgMemc->add( $redirKey, $url );
+		$this->getContext()->getOutput()->redirect( $url );
```

Comment:
I think it's worth putting `$this->getContext()->getOutput()` in a variable since it's used multiple times.